test_conformance/compiler/test_build_helpers.cpp

```diff
@@ -422,6 +443,118 @@ int test_get_program_info(cl_device_id deviceID, cl_context context, cl_command_
+                                       cl_context context,
+                                       cl_command_queue queue, int num_elements)
+{
+    int error = CL_SUCCESS;
+    size_t total_kernels = 0;
+    size_t kernel_names_len = 0;
+
+    clProgramWrapper program = nullptr;
+
+    // 1) Program without build call
+    //    Query CL_PROGRAM_NUM_KERNELS and check that it fails with
+    //    CL_INVALID_PROGRAM_EXECUTABLE Query CL_PROGRAM_KERNEL_NAMES and check
+    //    that it fails with CL_INVALID_PROGRAM_EXECUTABLE
+    {
+        program = clCreateProgramWithSource(
+            context, 1, &sample_multi_kernel_code_with_makro, nullptr, &error);
+        test_error(error, "clCreateProgramWithSource failed");
+
+        error = clGetProgramInfo(program, CL_PROGRAM_NUM_KERNELS,
+                                 sizeof(size_t), &total_kernels, nullptr);
+        test_failure_error(error, CL_INVALID_PROGRAM_EXECUTABLE,
+                           "Unexpected clGetProgramInfo result");
+
+        error = clGetProgramInfo(program, CL_PROGRAM_KERNEL_NAMES, 0, nullptr,
+                                 &kernel_names_len);
+        test_failure_error(error, CL_INVALID_PROGRAM_EXECUTABLE,
+                           "Unexpected clGetProgramInfo result");
+    }
+
+    // 2) Build the program with the preprocessor macro undefined
+    //    Query CL_PROGRAM_NUM_KERNELS and check that the correct number is
+    //    returned Query CL_PROGRAM_KERNEL_NAMES and check that the right kernel
+    //    names are returned
+    {
+        error =
+            clBuildProgram(program, 1, &deviceID, nullptr, nullptr, nullptr);
+        test_error(error, "clBuildProgram failed");
+
+        error = clGetProgramInfo(program, CL_PROGRAM_NUM_KERNELS,
+                                 sizeof(size_t), &total_kernels, nullptr);
+        test_error(error, "clGetProgramInfo failed");
+
+        test_assert_error(total_kernels == 2,
+                          "Unexpected clGetProgramInfo result");
+
+        error = clGetProgramInfo(program, CL_PROGRAM_KERNEL_NAMES, 0, nullptr,
+                                 &kernel_names_len);
+        test_error(error, "clGetProgramInfo failed");
+
+        std::vector<std::string> actual_names = { "sample_test_A",
+                                                  "sample_test_C" };
+
+        const size_t len = kernel_names_len + 1;
+        std::vector<char> kernel_names(len, '\0');
+        error =
+            clGetProgramInfo(program, CL_PROGRAM_KERNEL_NAMES, kernel_names_len,
+                             kernel_names.data(), &kernel_names_len);
+        test_error(error, "Unable to get kernel names list.");
+
+        std::string program_names = kernel_names.data();
+        for (const auto &name : actual_names)
+        {
+            test_assert_error(program_names.find(name.c_str())
+                                  != std::string::npos,
+                              "Unexpected kernel name");
+        }
```

Comment:
Should we add a check that the "forbidden" kernel name is not returned?  Something like
```
            test_assert_error(program_names.find("sample_test_B")
                                  == std::string::npos,
                              "sample_test_B should not be present");
```

Reply:
Thanks, that is indeed right thing to do - correction applied.